mlrun/package/packagers_manager.py

```diff
@@ -0,0 +1,654 @@
+                        # Link it (None will be used in case it was not found):
+                        artifact.spec.metrics[key] = joined_results.get(key, None)
+
+    def clear_packagers_outputs(self):
+        """
+        Clear the outputs of all packagers. This method should be called at the end of the run after logging all
+        artifacts as some will require uploading the files that will be deleted in this method.
+        """
+        for packager in self._packagers:
+            for path in packager.get_clearing_path_list():
+                if not os.path.exists(path):
+                    continue
+                if os.path.isdir(path):
+                    shutil.rmtree(path)
+                else:
+                    os.remove(path)
+
+    class _InstructionsNotesKey:
+        """
+        Library of keys for the packager instructions to be added to the packed artifact's spec.
+        """
+
+        PACKAGER_NAME = "packager_name"
+        OBJECT_TYPE = "object_type"
+        ARTIFACT_TYPE = "artifact_type"
+        INSTRUCTIONS = "instructions"
+
+    def _get_packager_by_name(self, name: str) -> Union[Type[Packager], None]:
+        """
+        Look for a packager with the given name and return it.
+
+        If a packager was not found None will be returned.
+
+        :param name: The name of the packager to get.
+
+        :return: The found packager or None if it wasn't found.
+        """
+        # Look for a packager by exact name:
+        for packager in self._packagers:
+            if packager.__name__ == name:
+                return packager
+
+        # No packager was found:
+        logger.warn(f"The packager '{name}' was not found.")
+        return None
+
+    def _get_packager_by_type(
+        self, object_type: Type, artifact_type: str = None
+    ) -> Union[Type[Packager], None]:
+        """
+        Look for a packager that can handle the provided object type and can also pack it as the provided artifact type.
+
+        If a packager was not found None will be returned.
+
+        :param object_type:   The object type the packager to get should handle.
+        :param artifact_type: The artifact type the packager to get should pack and log as.
+
+        :return: The found packager or None if it wasn't found.
+        """
+        # Look for a packager for the combination of object type nad artifact type:
+        possible_type_hints = [object_type]
+        while len(possible_type_hints) > 0:
+            for type_hint in possible_type_hints:
+                for packager in self._packagers:
+                    if packager.is_packable(
+                        object_type=type_hint, artifact_type=artifact_type
+                    ):
+                        return packager
+            # Reduce the type hint list and continue:
+            possible_type_hints = TypeHintUtils.reduce_type_hint(
+                type_hint=possible_type_hints
```

Comment:
As we discussed, your concern is totally justified, but the solutions suggested are not my favorite:
1. Agree it won't be a good thing.
2. IMO a packager should focus on a single type, not a list of types. If needed, a user can write a specific packager with the exact `Union[X, Y]` as the packing type. 

The more I thought about it the more I wanted to go back to my previous solution. We will try to unpack as all hinted types in the union and only if all unpacking failed we will raise an error. There are two scenarios where this would apply:
1. The user is using an old artifact or a path as its input, so it was not packaged and we don't have the original type - makes sense to simply try all options until we hit one (the user's code should support whatever was working as the type hint suggested is a union...) 
2. The user did used a packaged artifact, but he used a type hint that the original object type is not equal or part of (in case of a union). In this case, we warn the user and will try again to match whatever works.

WDYK?